core/vm/evm.go

```diff
@@ -602,6 +604,86 @@ func (evm *EVM) CallFromSystem(contractAddress common.Address, abi abipkg.ABI, f
 	return evm.handleABICall(abi, funcName, args, returnObj, call)
 }
 
+type cacheKey struct {
+	StateRoot       common.Hash
+	ContractAddress common.Address
+	TransactionData string
+}
+
+type cacheResult struct {
+	ret     []byte
+	gasLeft uint64
+}
+
+var staticCallCache, _ = lru.New(100)
+var (
+	cacheHits    = metrics.NewRegisteredMeter("contract_comm/caller/cache_hits", nil)
+	cacheMisses  = metrics.NewRegisteredMeter("contract_comm/caller/cache_misses", nil)
+	cacheSkipped = metrics.NewRegisteredMeter("contract_comm/caller/cache_skipped", nil)
+	cacheBadCast = metrics.NewRegisteredMeter("contract_comm/caller/cache_bad_cast", nil)
+	cacheCheck   = metrics.NewRegisteredMeter("contract_comm/caller/cache_check", nil)
+)
+
+func (evm *EVM) MemoizedStaticCallFromSystem(contractAddress common.Address, abi abipkg.ABI, funcName string, args []interface{}, returnObj interface{}, gas uint64) (uint64, error) {
+	staticCall := func(transactionData []byte) ([]byte, uint64, error) {
+		if !evm.StateDB.IsClean() || staticCallCache == nil {
+			cacheSkipped.Mark(1)
+			return evm.StaticCall(systemCaller, contractAddress, transactionData, gas)
+		}
+		stateRoot := evm.StateDB.StateRoot()
+
+		key := cacheKey{
+			TransactionData: string(transactionData),
+			ContractAddress: contractAddress,
+			StateRoot:       stateRoot,
+		}
+		if result, ok := staticCallCache.Get(key); ok {
+			cachedResult, castOk := result.(cacheResult)
+			if !castOk {
+				cacheBadCast.Mark(1)
+				staticCallCache.Remove(key)
+				return evm.StaticCall(systemCaller, contractAddress, transactionData, gas)
+			}
+			cacheHits.Mark(1)
+			ret := make([]byte, len(cachedResult.ret))
+			copy(ret, cachedResult.ret)
+			if evm.vmConfig.CheckStaticCallCache {
+				actualRet, actualGasLeft, actualErr := evm.StaticCall(systemCaller, contractAddress, transactionData, gas)
+				if actualErr != nil {
+					log.Error("non nil error when performing evm static call that wanted to use the cached result")
```

Comment:
We should make sure these error messages have a string that makes them easy to find for error detection purposes. "cached result" is probably good enough, but "static call cached result" might be better.